### PR TITLE
Add logic to skip files with more than 15000 lines

### PR DESCRIPTION
**Pull request by ariskataoka**
TLDR; Some builds are currently timing out due to PRs that contain too large files. 

This PR addresses [issue 147](https://github.com/Automattic/vip-go-ci/issues/147).

If the PR contains files with more than 15000 lines, the lint scan and the phpcs will skip it. 

- Adds logic to validate file size;
- Adds logic to modify the bot message;
- Adds logic to ignore files that have been skipped previously;

TODO:
- [ X ] Implement [new feature / logic]
- [ X ] Add/update unit-tests
- [ X ] Manual testing
 - [ X ] Pull-Request with files with more than 15000 lines

Obs.: 

1. Tested manually with lint and php-cs on and off: https://github.com/ariskataoka/vipgoci-tests/pull/10#pullrequestreview-697919089

2. The build testing has constantly been failing due to the GH API limit rate.
```
[ 2021-05-26T08:37:48+00:00 -- 0 ]  Ran out of request limits for GitHub, cannot continue without making further requests.; {
    "github_url": "https:\/\/api.github.com\/repos\/automattic\/vip-go-ci-testing\/pulls\/39\/commits?page=1&per_page=100",
    "x-ratelimit-remaining": "1",
    "x-ratelimit-limit": "60"
}
```

2. Original branch/PR here: https://github.com/Automattic/vip-go-ci/pull/176

### CODE REVIEW

File: defines.php [hunk 1]
```diff
@@ -134,7 +134,23 @@
...
 define( 'VIPGOCI_GIT_DIFF_DATA_SOURCE_GITHUB_API',	'github-api' );
 
+
+
+/**
+ * Define file number of lines limit
+ */
+define( 'VIPGOCI_SKIPPED_FILES',	        'skipped-files'	);
+
+define( 'VIPGOCI_VALIDATION_MAXIMUM_LINES_LIMIT',   15000   );
```
Comment (gudmdharalds) [inline defines.php:148]: Would be best if this was configurable via command-line. Suggest that the option is named `--skip-large-files-maxlines` or similar.

File: tests/A09LintLintScanCommitTest.php [hunk 2]
```diff
@@ -324,5 +335,103 @@ public function testLintDoScan2() {
...
+			)
+		);
+
+		$this->assertSame(
+			$expected_issues_skipped,
+			$issues_skipped
+		);
+
+		unset( $this->options['commit'] );
+	}
```
Comment (gudmdharalds) [inline tests/A09LintLintScanCommitTest.php:428]: Looks good, but suggest to add test for a different condition: When `skip-large-files` is set to `false` and there is a large file in the commit scanned.

File: tests/PhpcsScanScanCommitTest.php [hunk 3]
```diff
@@ -896,4 +924,98 @@ public function testDoScanTest5() {
...
+			$issues_stats[
+			$pr_item->number
+			][
+			'error'
+			] = 0;
+
+			$issues_skipped[ $pr_item->number ] = $this->getDefaultSkippedFilesDueIssuesMock();
+		}
+
+		$issues_skipped[39] = array(
```
Comment (gudmdharalds) [inline tests/PhpcsScanScanCommitTest.php:970]: Should this not be set by the functions called by `vipgoci_phpcs_scan_commit()`? Tested running the unit-test without this and it succeeded. Seems like that this code could be removed.

File: tests/PhpcsScanScanCommitTest.php [hunk 4]
```diff
@@ -896,4 +924,98 @@ public function testDoScanTest5() {
...
+				'total' => 1
+			)
+		);
+
+		$this->assertSame(
+			$expected_issues_skipped[39],
+			$issues_skipped[39]
+		);
+	}
+
```
Comment (gudmdharalds) [inline tests/PhpcsScanScanCommitTest.php:1014]: Same here, would be good to have a test with large file existing and `skip-large-files` set to `false`.

File: tests/VipgociSkipFileTest.php [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+		$skipped_files = vipgoci_get_skipped_files(
+			$current_skipped_files_mock,
+			$validation_mock
+		);
+
+		$expected_skipped_files = array(
+			'total' => 3,
+			'issues' => array(
+                'max-lines' => array(
```
Comment (gudmdharalds) [inline tests/VipgociSkipFileTest.php:36]: Missing tabbed indenting, spaces used.

**Review by gudmdharalds — CHANGES_REQUESTED** [review]
This looks very good! 

I left only a few comments inline.

File: tests/PhpcsScanScanCommitTest.php [hunk 6]
```diff
@@ -896,4 +924,98 @@ public function testDoScanTest5() {
...
+		$expected_issues_skipped = array(
+			39 => array(
+				'issues' => array(
+					'max-lines' => array ( 'test1/myfile-1.php' )
+				),
+				'total' => 1
+			)
+		);
+
```
Comment (gudmdharalds) [inline tests/PhpcsScanScanCommitTest.php:1008]: Suggest to a) add a new file to the commit that would be scanned and has issues, b) to add some code with problems to the file that is skipped, and c) and then check here if `$issues_submit` contains only issues for the first file. This way we know that PHPCS scanning returns with issues even if some files were skipped, and we know that that they are in the correct file.

This would make the test a bit more comprehensive.

**Review by gudmdharalds — CHANGES_REQUESTED** [review]
Hi,

I reviewed and this PR is looking excellent. 

There is just one small issue I noted, and when that is resolved this PR can be merged.

Can you have a look?

Thanks!

**Review by gudmdharalds — APPROVED** [review]
